# EB5202 Web Analytics - Retail Rocket - Feature Engineering

## Load libraries

In [39]:
import os
import pandas as pd
import pickle
from collections import Counter

## Load data

In [2]:
sessions_df = pd.read_pickle(os.path.join('data','sessions.pkl'))

In [3]:
sessions_df.head()

<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>visitorid</th>
      <th>startdate</th>
      <th>enddate</th>
      <th>addtocart</th>
      <th>transaction</th>
      <th>view</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>0</td>
      <td>2015-09-11 20:49:49.439</td>
      <td>2015-09-11 20:55:17.175</td>
      <td>[]</td>
      <td>[]</td>
      <td>[285930, 357564, 67045]</td>
    </tr>
    <tr>
      <th>1</th>
      <td>1</td>
      <td>2015-08-13 17:46:06.444</td>
      <td>2015-08-13 17:46:06.444</td>
      <td>[]</td>
      <td>[]</td>
      <td>[72028]</td>
    </tr>
    <tr>
      <th>2</th>
      <td>2</td>
      <td>2015-08-07 17:51:44.567</td>
      <td>2015-08-07 18:20:57.845</td>
      <td>[]</td>
      <td>[]</td>
      <td>[325215, 325215, 259884, 216305, 342816, 34281...</td>
    </tr>
    <tr>
      <th>3</th>
      <td>3</td>
      <td>2015-08-01 07:10:35.296</td>
      <td>2015-08-01 07:10:35.296</td>
      <td>[]</td>
      <td>[]</td>
      <td>[385090]</td>
    </tr>
    <tr>
      <th>4</th>
      <td>4</td>
      <td>2015-09-15 21:24:27.167</td>
      <td>2015-09-15 21:24:27.167</td>
      <td>[]</td>
      <td>[]</td>
      <td>[177677]</td>
    </tr>
  </tbody>
</table>
</div>

## Create features

### Pages

In [5]:
sessions_df['pages'] = sessions_df['view'].apply(lambda x: len(x))
sessions_df.head()

<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>visitorid</th>
      <th>startdate</th>
      <th>enddate</th>
      <th>addtocart</th>
      <th>transaction</th>
      <th>view</th>
      <th>pages</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>0</td>
      <td>2015-09-11 20:49:49.439</td>
      <td>2015-09-11 20:55:17.175</td>
      <td>[]</td>
      <td>[]</td>
      <td>[285930, 357564, 67045]</td>
      <td>3</td>
    </tr>
    <tr>
      <th>1</th>
      <td>1</td>
      <td>2015-08-13 17:46:06.444</td>
      <td>2015-08-13 17:46:06.444</td>
      <td>[]</td>
      <td>[]</td>
      <td>[72028]</td>
      <td>1</td>
    </tr>
    <tr>
      <th>2</th>
      <td>2</td>
      <td>2015-08-07 17:51:44.567</td>
      <td>2015-08-07 18:20:57.845</td>
      <td>[]</td>
      <td>[]</td>
      <td>[325215, 325215, 259884, 216305, 342816, 34281...</td>
      <td>8</td>
    </tr>
    <tr>
      <th>3</th>
      <td>3</td>
      <td>2015-08-01 07:10:35.296</td>
      <td>2015-08-01 07:10:35.296</td>
      <td>[]</td>
      <td>[]</td>
      <td>[385090]</td>
      <td>1</td>
    </tr>
    <tr>
      <th>4</th>
      <td>4</td>
      <td>2015-09-15 21:24:27.167</td>
      <td>2015-09-15 21:24:27.167</td>
      <td>[]</td>
      <td>[]</td>
      <td>[177677]</td>
      <td>1</td>
    </tr>
  </tbody>
</table>
</div>

In [11]:
(sessions_df.loc[0, 'enddate'] - sessions_df.loc[0, 'startdate']) / sessions_df.loc[0, 'pages']

numpy.timedelta64(109245333333,'ns')

In [19]:
sessions_df['pages'].describe()

count    1.786340e+06
mean     1.491492e+00
std      1.591950e+00
min      0.000000e+00
25%      1.000000e+00
50%      1.000000e+00
75%      1.000000e+00
max      1.900000e+02
Name: pages, dtype: float64

In [20]:
sessions_df.loc[sessions_df['pages'] == 0].head()

<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>visitorid</th>
      <th>startdate</th>
      <th>enddate</th>
      <th>addtocart</th>
      <th>transaction</th>
      <th>view</th>
      <th>pages</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>6</th>
      <td>6</td>
      <td>2015-08-30 06:03:48.202</td>
      <td>2015-08-30 06:03:48.202</td>
      <td>[65273]</td>
      <td>[]</td>
      <td>[]</td>
      <td>0</td>
    </tr>
    <tr>
      <th>211</th>
      <td>155</td>
      <td>2015-09-11 06:12:12.678</td>
      <td>2015-09-11 06:19:00.395</td>
      <td>[368372, 452082, 181405, 41882, 442601, 224623]</td>
      <td>[]</td>
      <td>[]</td>
      <td>0</td>
    </tr>
    <tr>
      <th>212</th>
      <td>155</td>
      <td>2015-09-12 04:38:00.026</td>
      <td>2015-09-11 06:19:00.395</td>
      <td>[389974]</td>
      <td>[]</td>
      <td>[]</td>
      <td>0</td>
    </tr>
    <tr>
      <th>256</th>
      <td>186</td>
      <td>2015-08-12 16:28:16.523</td>
      <td>2015-08-12 16:34:57.040</td>
      <td>[49029]</td>
      <td>[49029]</td>
      <td>[]</td>
      <td>0</td>
    </tr>
    <tr>
      <th>527</th>
      <td>419</td>
      <td>2015-07-29 05:03:12.695</td>
      <td>2015-07-29 04:03:49.136</td>
      <td>[]</td>
      <td>[19278]</td>
      <td>[]</td>
      <td>0</td>
    </tr>
  </tbody>
</table>
</div>

### Pagetime

In [35]:
pages_more_than1 = sessions_df['pages'] > 1    

In [36]:
sessions_df.loc[pages_more_than1, 'pagetime'] = (sessions_df.loc[pages_more_than1, 'enddate'] - sessions_df.loc[pages_more_than1, 'startdate']) /\
                                                (sessions_df.loc[pages_more_than1, 'pages'] - 1)

In [37]:
pages_less_than1 = pages_more_than1.apply(lambda x: not x)

In [38]:
sessions_df.loc[pages_less_than1, 'pagetime'] = pd.Timedelta(0)

In [39]:
sessions_df['pagetime'].to_frame().info(null_counts=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1786340 entries, 0 to 1786339
Data columns (total 1 columns):
pagetime    1786340 non-null timedelta64[ns]
dtypes: timedelta64[ns](1)
memory usage: 13.6 MB


In [19]:
sessions_df.head(10)

<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>visitorid</th>
      <th>startdate</th>
      <th>enddate</th>
      <th>addtocart</th>
      <th>transaction</th>
      <th>view</th>
      <th>pages</th>
      <th>pagetime</th>
      <th>diffcat</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>0</td>
      <td>2015-09-11 20:49:49.439</td>
      <td>2015-09-11 20:55:17.175</td>
      <td>[]</td>
      <td>[]</td>
      <td>[285930, 357564, 67045]</td>
      <td>3</td>
      <td>00:02:43.868000</td>
      <td>1.00</td>
    </tr>
    <tr>
      <th>1</th>
      <td>1</td>
      <td>2015-08-13 17:46:06.444</td>
      <td>2015-08-13 17:46:06.444</td>
      <td>[]</td>
      <td>[]</td>
      <td>[72028]</td>
      <td>1</td>
      <td>00:00:00</td>
      <td>1.00</td>
    </tr>
    <tr>
      <th>2</th>
      <td>2</td>
      <td>2015-08-07 17:51:44.567</td>
      <td>2015-08-07 18:20:57.845</td>
      <td>[]</td>
      <td>[]</td>
      <td>[325215, 325215, 259884, 216305, 342816, 34281...</td>
      <td>8</td>
      <td>00:04:10.468285</td>
      <td>0.25</td>
    </tr>
    <tr>
      <th>3</th>
      <td>3</td>
      <td>2015-08-01 07:10:35.296</td>
      <td>2015-08-01 07:10:35.296</td>
      <td>[]</td>
      <td>[]</td>
      <td>[385090]</td>
      <td>1</td>
      <td>00:00:00</td>
      <td>1.00</td>
    </tr>
    <tr>
      <th>4</th>
      <td>4</td>
      <td>2015-09-15 21:24:27.167</td>
      <td>2015-09-15 21:24:27.167</td>
      <td>[]</td>
      <td>[]</td>
      <td>[177677]</td>
      <td>1</td>
      <td>00:00:00</td>
      <td>1.00</td>
    </tr>
    <tr>
      <th>5</th>
      <td>5</td>
      <td>2015-07-17 01:45:56.439</td>
      <td>2015-07-17 01:45:56.439</td>
      <td>[]</td>
      <td>[]</td>
      <td>[61396]</td>
      <td>1</td>
      <td>00:00:00</td>
      <td>1.00</td>
    </tr>
    <tr>
      <th>6</th>
      <td>6</td>
      <td>2015-08-30 06:03:48.202</td>
      <td>2015-08-30 06:03:48.202</td>
      <td>[65273]</td>
      <td>[]</td>
      <td>[]</td>
      <td>0</td>
      <td>00:00:00</td>
      <td>0.00</td>
    </tr>
    <tr>
      <th>7</th>
      <td>6</td>
      <td>2015-08-30 06:39:38.318</td>
      <td>2015-08-30 06:56:33.838</td>
      <td>[]</td>
      <td>[]</td>
      <td>[253615, 344723, 344723, 344723]</td>
      <td>4</td>
      <td>00:05:38.506666</td>
      <td>0.25</td>
    </tr>
    <tr>
      <th>8</th>
      <td>6</td>
      <td>2015-08-31 03:21:25.697</td>
      <td>2015-08-30 06:56:33.838</td>
      <td>[]</td>
      <td>[]</td>
      <td>[344723]</td>
      <td>1</td>
      <td>00:00:00</td>
      <td>1.00</td>
    </tr>
    <tr>
      <th>9</th>
      <td>7</td>
      <td>2015-05-14 05:39:36.753</td>
      <td>2015-05-14 05:42:42.817</td>
      <td>[]</td>
      <td>[]</td>
      <td>[139394, 164941]</td>
      <td>2</td>
      <td>00:03:06.064000</td>
      <td>0.50</td>
    </tr>
  </tbody>
</table>
</div>

Note that there are vistorid (e.g. 6) with multiple sessions (visits).

### Diffcat

In [2]:
infile = open(os.path.join('data', 'itemtocat.pkl'), 'rb')
itemtocat_dict = pickle.load(infile)
infile.close()

In [69]:
itemtocat_dict[177677]

KeyError: 177677

There are itemids which are not found in the dictionary. Need to handle this when converting itemid to categoryid.

In [18]:
sessions_df.loc[10, :]

visitorid                               7
startdate      2015-05-16 04:20:39.214000
enddate        2015-05-14 05:42:42.817000
addtocart                              []
transaction                            []
view                             [226353]
pages                                   1
pagetime                  0 days 00:00:00
diffcat                                 1
Name: 10, dtype: object

In [15]:
for i in sessions_df['view'][10]:
    print(itemtocat_dict[i])

[(Timestamp('2015-08-02 03:00:00'), '642'), (Timestamp('2015-09-13 03:00:00'), '429')]


In [3]:
def to_categoryid(itemid, enddate):
    catid = ''
    
    for time, cat in itemtocat_dict.get(itemid, []):
        if enddate <= time:
            catid = cat
            break
    if not catid:
        if itemtocat_dict.get(itemid, []):  # enddate did not match
            catid = itemtocat_dict[itemid][-1][1]  # set categiryid to last value
        else:
            catid = str(itemid) + '_no_cat_id'  # itemid not in dictionary
    return catid

In [20]:
{ to_categoryid(i, sessions_df['enddate'][10]) for i in sessions_df['view'][10] }

{'642'}

Given the session enddate of 2015-05-14, the categoryid 642 is correct.

In [74]:
sessions_df['diffcat'] = sessions_df.loc[:, ['enddate', 'view']].apply(
    lambda x: len({ to_categoryid(i, x['enddate']) for i in x['view'] }) / len(x['view']) if x['view'] else 0, axis=1)
#sessions_df['diffcat'] = sessions_df.loc[:, ['enddate', 'view']].apply(lambda x: (x.view), axis=1)
#sessions_df.loc[[0, 1, 2], ['enddate', 'view']].apply(lambda x: print(x['view']), axis=1)

In [75]:
sessions_df.head()

<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>visitorid</th>
      <th>startdate</th>
      <th>enddate</th>
      <th>addtocart</th>
      <th>transaction</th>
      <th>view</th>
      <th>pages</th>
      <th>pagetime</th>
      <th>diffcat</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>0</td>
      <td>2015-09-11 20:49:49.439</td>
      <td>2015-09-11 20:55:17.175</td>
      <td>[]</td>
      <td>[]</td>
      <td>[285930, 357564, 67045]</td>
      <td>3</td>
      <td>00:02:43.868000</td>
      <td>1.00</td>
    </tr>
    <tr>
      <th>1</th>
      <td>1</td>
      <td>2015-08-13 17:46:06.444</td>
      <td>2015-08-13 17:46:06.444</td>
      <td>[]</td>
      <td>[]</td>
      <td>[72028]</td>
      <td>1</td>
      <td>00:00:00</td>
      <td>1.00</td>
    </tr>
    <tr>
      <th>2</th>
      <td>2</td>
      <td>2015-08-07 17:51:44.567</td>
      <td>2015-08-07 18:20:57.845</td>
      <td>[]</td>
      <td>[]</td>
      <td>[325215, 325215, 259884, 216305, 342816, 34281...</td>
      <td>8</td>
      <td>00:04:10.468285</td>
      <td>0.25</td>
    </tr>
    <tr>
      <th>3</th>
      <td>3</td>
      <td>2015-08-01 07:10:35.296</td>
      <td>2015-08-01 07:10:35.296</td>
      <td>[]</td>
      <td>[]</td>
      <td>[385090]</td>
      <td>1</td>
      <td>00:00:00</td>
      <td>1.00</td>
    </tr>
    <tr>
      <th>4</th>
      <td>4</td>
      <td>2015-09-15 21:24:27.167</td>
      <td>2015-09-15 21:24:27.167</td>
      <td>[]</td>
      <td>[]</td>
      <td>[177677]</td>
      <td>1</td>
      <td>00:00:00</td>
      <td>1.00</td>
    </tr>
  </tbody>
</table>
</div>

In [28]:
sessions_df['view'][2]

[325215, 325215, 259884, 216305, 342816, 342816, 216305, 325215]

In [29]:
{ to_categoryid(i, sessions_df['enddate'][2]) for i in sessions_df['view'][2] }

{'299', '444'}

diffcat is 2 / 8 = .25 which is correct in the above table.

In [79]:
sessions_df['diffcat'].to_frame().info(null_counts=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1786340 entries, 0 to 1786339
Data columns (total 1 columns):
diffcat    1786340 non-null float64
dtypes: float64(1)
memory usage: 13.6 MB


In [80]:
sessions_df['diffcat'].describe()

count    1.786340e+06
mean     8.866262e-01
std      2.385307e-01
min      0.000000e+00
25%      1.000000e+00
50%      1.000000e+00
75%      1.000000e+00
max      1.000000e+00
Name: diffcat, dtype: float64

### Diffprod

In [23]:
sessions_df['diffprod'] = sessions_df['view'].map(lambda x: len(set(x)) / len(x) if x else 0)

In [24]:
sessions_df.head()

<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>visitorid</th>
      <th>startdate</th>
      <th>enddate</th>
      <th>addtocart</th>
      <th>transaction</th>
      <th>view</th>
      <th>pages</th>
      <th>pagetime</th>
      <th>diffcat</th>
      <th>diffprod</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>0</td>
      <td>2015-09-11 20:49:49.439</td>
      <td>2015-09-11 20:55:17.175</td>
      <td>[]</td>
      <td>[]</td>
      <td>[285930, 357564, 67045]</td>
      <td>3</td>
      <td>00:02:43.868000</td>
      <td>1.00</td>
      <td>1.0</td>
    </tr>
    <tr>
      <th>1</th>
      <td>1</td>
      <td>2015-08-13 17:46:06.444</td>
      <td>2015-08-13 17:46:06.444</td>
      <td>[]</td>
      <td>[]</td>
      <td>[72028]</td>
      <td>1</td>
      <td>00:00:00</td>
      <td>1.00</td>
      <td>1.0</td>
    </tr>
    <tr>
      <th>2</th>
      <td>2</td>
      <td>2015-08-07 17:51:44.567</td>
      <td>2015-08-07 18:20:57.845</td>
      <td>[]</td>
      <td>[]</td>
      <td>[325215, 325215, 259884, 216305, 342816, 34281...</td>
      <td>8</td>
      <td>00:04:10.468285</td>
      <td>0.25</td>
      <td>0.5</td>
    </tr>
    <tr>
      <th>3</th>
      <td>3</td>
      <td>2015-08-01 07:10:35.296</td>
      <td>2015-08-01 07:10:35.296</td>
      <td>[]</td>
      <td>[]</td>
      <td>[385090]</td>
      <td>1</td>
      <td>00:00:00</td>
      <td>1.00</td>
      <td>1.0</td>
    </tr>
    <tr>
      <th>4</th>
      <td>4</td>
      <td>2015-09-15 21:24:27.167</td>
      <td>2015-09-15 21:24:27.167</td>
      <td>[]</td>
      <td>[]</td>
      <td>[177677]</td>
      <td>1</td>
      <td>00:00:00</td>
      <td>1.00</td>
      <td>1.0</td>
    </tr>
  </tbody>
</table>
</div>

In [25]:
sessions_df['view'][2]

[325215, 325215, 259884, 216305, 342816, 342816, 216305, 325215]

diffprod is 4 / 8 = .5

In [35]:
sessions_df['diffprod'].to_frame().info(null_counts=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1786340 entries, 0 to 1786339
Data columns (total 1 columns):
diffprod    1786340 non-null float64
dtypes: float64(1)
memory usage: 13.6 MB


In [36]:
sessions_df['diffprod'].describe()

count    1.786340e+06
mean     9.455984e-01
std      1.600542e-01
min      0.000000e+00
25%      1.000000e+00
50%      1.000000e+00
75%      1.000000e+00
max      1.000000e+00
Name: diffprod, dtype: float64

### Prodcat

In [32]:
sessions_df['prodcat'] = sessions_df.loc[:, ['enddate', 'view']].apply(
    lambda x: len(set(x['view'])) / len({ to_categoryid(i, x['enddate']) for i in x['view'] }) if x['view'] else 0, axis=1)

In [33]:
sessions_df.head()

<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>visitorid</th>
      <th>startdate</th>
      <th>enddate</th>
      <th>addtocart</th>
      <th>transaction</th>
      <th>view</th>
      <th>pages</th>
      <th>pagetime</th>
      <th>diffcat</th>
      <th>diffprod</th>
      <th>prodcat</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>0</td>
      <td>2015-09-11 20:49:49.439</td>
      <td>2015-09-11 20:55:17.175</td>
      <td>[]</td>
      <td>[]</td>
      <td>[285930, 357564, 67045]</td>
      <td>3</td>
      <td>00:02:43.868000</td>
      <td>1.00</td>
      <td>1.0</td>
      <td>1.0</td>
    </tr>
    <tr>
      <th>1</th>
      <td>1</td>
      <td>2015-08-13 17:46:06.444</td>
      <td>2015-08-13 17:46:06.444</td>
      <td>[]</td>
      <td>[]</td>
      <td>[72028]</td>
      <td>1</td>
      <td>00:00:00</td>
      <td>1.00</td>
      <td>1.0</td>
      <td>1.0</td>
    </tr>
    <tr>
      <th>2</th>
      <td>2</td>
      <td>2015-08-07 17:51:44.567</td>
      <td>2015-08-07 18:20:57.845</td>
      <td>[]</td>
      <td>[]</td>
      <td>[325215, 325215, 259884, 216305, 342816, 34281...</td>
      <td>8</td>
      <td>00:04:10.468285</td>
      <td>0.25</td>
      <td>0.5</td>
      <td>2.0</td>
    </tr>
    <tr>
      <th>3</th>
      <td>3</td>
      <td>2015-08-01 07:10:35.296</td>
      <td>2015-08-01 07:10:35.296</td>
      <td>[]</td>
      <td>[]</td>
      <td>[385090]</td>
      <td>1</td>
      <td>00:00:00</td>
      <td>1.00</td>
      <td>1.0</td>
      <td>1.0</td>
    </tr>
    <tr>
      <th>4</th>
      <td>4</td>
      <td>2015-09-15 21:24:27.167</td>
      <td>2015-09-15 21:24:27.167</td>
      <td>[]</td>
      <td>[]</td>
      <td>[177677]</td>
      <td>1</td>
      <td>00:00:00</td>
      <td>1.00</td>
      <td>1.0</td>
      <td>1.0</td>
    </tr>
  </tbody>
</table>
</div>

For visitorid 2, previously unique categoryid was determined to be 2 (see diffcat), the unique itemid was 4. prodcat is 4 / 2 = 2 which is correct in the above table.

In [38]:
sessions_df['prodcat'].to_frame().info(null_counts=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1786340 entries, 0 to 1786339
Data columns (total 1 columns):
prodcat    1786340 non-null float64
dtypes: float64(1)
memory usage: 13.6 MB


In [37]:
sessions_df['prodcat'].describe()

count    1.786340e+06
mean     1.184270e+00
std      7.023515e-01
min      0.000000e+00
25%      1.000000e+00
50%      1.000000e+00
75%      1.000000e+00
max      5.400000e+01
Name: prodcat, dtype: float64

### Maxrep

In [40]:
sessions_df['maxrep'] = sessions_df['view'].map(lambda x: Counter(x).most_common(1)[0][1] if x else 0)

In [44]:
sessions_df.iloc[:5, -8:]

<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>transaction</th>
      <th>view</th>
      <th>pages</th>
      <th>pagetime</th>
      <th>diffcat</th>
      <th>diffprod</th>
      <th>prodcat</th>
      <th>maxrep</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>[]</td>
      <td>[285930, 357564, 67045]</td>
      <td>3</td>
      <td>00:02:43.868000</td>
      <td>1.00</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>1</td>
    </tr>
    <tr>
      <th>1</th>
      <td>[]</td>
      <td>[72028]</td>
      <td>1</td>
      <td>00:00:00</td>
      <td>1.00</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>1</td>
    </tr>
    <tr>
      <th>2</th>
      <td>[]</td>
      <td>[325215, 325215, 259884, 216305, 342816, 34281...</td>
      <td>8</td>
      <td>00:04:10.468285</td>
      <td>0.25</td>
      <td>0.5</td>
      <td>2.0</td>
      <td>3</td>
    </tr>
    <tr>
      <th>3</th>
      <td>[]</td>
      <td>[385090]</td>
      <td>1</td>
      <td>00:00:00</td>
      <td>1.00</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>1</td>
    </tr>
    <tr>
      <th>4</th>
      <td>[]</td>
      <td>[177677]</td>
      <td>1</td>
      <td>00:00:00</td>
      <td>1.00</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>1</td>
    </tr>
  </tbody>
</table>
</div>

In [45]:
sessions_df['view'][2]

[325215, 325215, 259884, 216305, 342816, 342816, 216305, 325215]

itemid 325215 was viewed 3 times, maxrep = 3 is correct in the above table.

## Save new features to file

In [81]:
sessions_df.to_pickle(os.path.join('data', 'sessions_newfeatures.pkl'))
# sessions_df = pd.read_pickle(os.path.join('data', 'sessions_temp.pkl'))

## Split data into train / test dataset

In [21]:
# train_idx = random.sample(range(len(sessions_df)), int(len(sessions_df) / 2))
# test_idx = [ i for i in range(len(sessions_df)) if i not in train_idx ]